[Code Refactoring] Clean up `Constructors` that take number or BigInt

thelucif3r opened the pull request:
1. Add 3 constant variables to avoid typos: MAX_LENGTH, ALL_ONES, ONE_ALL_ZEROS.
2. Refactor 3 constructors that take long, ulong, and BigInteger and the default one.

<!---
@huboard:{"custom_state":"archived"}
-->

8mpt committed:
[Code Refactoring] Clean up `Constructors` that take number or BigInteger.

1. Add 3 constant variables to avoid typos: MAX_LENGTH, ALL_ONES, ONE_ALL_ZEROS.
2. Refactor 3 constructors that take long, ulong, and BigInteger and the default one.

bazzilic commented on BigInteger/BigInteger.cs:
```diff
@@ -5,6 +5,8 @@ public class BigInteger
     // maximum length of the BigInteger in uint (4 bytes)
     // change this to suit the required level of precision.
     private const int maxLength = 70;
+    private const int MAX_LENGTH = 70;
```
What is the purpose of `MAX_LENGTH` as compared to `maxLength`?

thelucif3r replied:
MAX_LENGTH is there to replace maxLength. All constants' name should follow conventions to increase code readability.

bazzilic commented on BigInteger/BigInteger.cs:
```diff
@@ -26,77 +28,49 @@ public class BigInteger
-    public int dataLength;                 // number of actual chars used
-
+    private uint[] data = new uint[MAX_LENGTH];             // stores bytes from the Big Integer
+    public int dataLength { get; set; }                 // number of actual chars used
```
Is there a reason to change the field to property?

thelucif3r replied:
I made a mistake here to put `setter` there because I saw it used `public`. However, is it meant to be `private`?

bazzilic commented on BigInteger/BigInteger.cs:
```diff
@@ -26,77 +28,49 @@ public class BigInteger
-        if (dataLength == 0)
-            dataLength = 1;
-    }
+    public BigInteger(ulong value) : this((long)value) { }
```
Here's the reason why we need tests. What would happen in this code?

``` csharp
ulong x = 0xDEADBEEFDEADBEEF;
var bi = new BigInteger(x);
```

thelucif3r replied:
Okay, I will add some test cases to `BigIntegerTest/Constructors.cs`.

bazzilic commented:
Changes that might affect behavior can only be accepted after the behavior is covered with tests.
Also, the internals of the class will undergo massive refactoring so that it could be variable length.

8mpt committed:
[Fix Previous Commit] reimplement ctor(ulong) with some code refactoring

Fix the previous commit which had some misunderstanding on `shift` operator and type conversion.

bazzilic commented:
This PR is a lot of unnecessary changes in the code that is known to work.
Not merging it in.
